source/transfer_list.rst

```diff
@@ -146,6 +146,224 @@ The TE header is defined in :numref:`tab_te_header`.
      - The exact size of the data content in bytes, not including inter-TE padding. May be 0.
 
 
+.. _sec_operations:
+
+Standard operations
+-------------------
+
+This section describes the valid operations that can be performed on a TL in
+more detail, in order to clarify how to use the various fields and to serve as a
+guideline for implementation.
+
+Validating a TL header
+^^^^^^^^^^^^^^^^^^^^^^
+
+.. default-role:: code
+
+Inputs:
+
+- `tl_base_addr`: Base address of the existing TL.
+
+#. Compare `tl.signature` (`tl_base_addr + 0x0`) to `0x6e_d0ff`. On a mismatch,
+   abort (this is not a valid TL).
+
+#. Compare `tl.version` (`tl_base_addr + 0x5`) to the expected version
+   (currently |current_version|). If there is an exact match, the TL is valid
+   for all operations outlined in this section. If `tl.version` is larger, the
+   TL is valid for reading but must not be modified or relocated. If
```

Comment:
Sorry, I know this is now merged but I'd just like to question this statement. If we can assume that all future versions will be backwards compatible, then why is not safe to modify or relocate the TL according to this version of the spec? Not being able to at least relocate a TL could be problematic, for example when passing a TL through a component that has not upgraded to the latest version but needs to make a deep copy (e.g. due to the original memory location being inaccessible to a subsequent component).

Reply:
It is true that this could be quite a large limitation, but increasing the version may introduce a new restriction or feature that the old software cannot support.

The implication here is that the software needs to be updated. I can certainly see a problem here though, since it means that an update of an earlier component can break later phases.

Reply:
I... huh. Strange. I thought we had agreed that it would be backwards-compatible only for reading but not for writing (I don't think it was my idea), but I can't find any source for that anymore. Maybe I just totally misread your comment here (which actually says the opposite): https://github.com/FirmwareHandoff/firmware_handoff/pull/4#discussion_r1040852918

This is a trade-off that can go either way: if we say that older implementations are still allowed to modify new TLs, then that limits the amount of things we can add to the spec as a backwards-compatible update later. For example, if for whatever reason we one day decide that in addition to the simple checksum the TL header should now contain a SHA256 of the whole TL, that would be fine as a backwards-compatible update if older implementations have to treat the TL as read-only, but it would break if they're also allowed to append (because then they wouldn't know to update that hash). Of course, on the other hand not being able to append limits the amount of useful interop you can have with older implementations.

I don't have a strong opinion on this either way. If you say you prefer that older implementations be allowed to write as well, I'm happy to write another PR to change that. We could even split the version number into 4 and 4 bits if we want to potentially allow both (although that starts getting a bit complicated).

Allowing only relocation and reading is also possible, that just means you cannot make any changes in newer versions that would potentially affect relocation behavior (e.g. if we wanted to add a "relocation_count" field that counts how often the TL was relocated for some reason). It's probably pretty unlikely that that would ever come up but it's hard to predict future requirements perfectly.

Reply:
Yes, you're (both) right it's not in general safe for old code to modify new TLs so I don't think we can change this policy in isolation. I wondered about using some bits of the version field to indicate to old code whether it's safe to modify the TL, but it seems unlikely that a new TL/TE header field will be added without affecting old writers in some way.

I'm still concerned about the interop problem but hopefully incrementing the version is sufficiently rare to make this not a problem in practice. I suppose old code could recreate the TL in the format it understands and hope that critical information is not lost when handing off to the next layer.